### DISCUSSION TITLE
Transform Http requests to https requests with certificate

### DISCUSSION THREAD

**Discussion by codebased** [Kong/kong#8485] [Help]
Hi, How can I translate the http request to https through a plugin ?

Background: 
I have HTTP requests coming with a header that has client certificate. I am running Kong with MTLS enabled. 
What I want is to transform a request to convert http request to https and attach certificate from header to body for MTLS plugin to validate.

**Comment by fffonion** [comment]
Hi @codebased, instead of thinking of converting http to https, you can implemente your own plugin that get the certificate from request header, and reuse the MTLS plugin's validation part of code to validate. As MTLS plugin is an enterprise-only feature, please open a support ticket if you need help on implementing your custom plugin.

**Reply by codebased** [reply]
Thanks @fffonion Yes we are on enterprise license. I will have to find who can create a support ticket because it is a big company and it may take time to find who has got the login. What is involved in writing a plugin to read from the header and reuse mtls plugin ?

**Reply by fffonion** [reply]
@codebased It's tricky to discuss here since I can't point you to the code, which is not open sourced. In the mtls-auth plugin you will see it calls these two API [resty.kong.tls.request_client_certificate](https://github.com/Kong/lua-kong-nginx-module#restykongtlsrequest_client_certificate) and [resty.kong.tls.get_full_client_certificate_chain](https://github.com/Kong/lua-kong-nginx-module#restykongtlsget_full_client_certificate_chain), you will need to remove the first one, and replace the second one with your own logic to get it from header. All other part of code can remain unchanged.
